pkg/local_object_storage/metabase/version.go

```diff
@@ -112,83 +115,118 @@ func migrateFrom2Version(db *DB, tx *bbolt.Tx) error {
+			var err error
+			if fromBkt, afterObj, err = migrateContainersToMetaBucket(l, cs, tx, fromBkt, afterObj); err == nil {
+				fromBkt, afterObj = slices.Clone(fromBkt), slices.Clone(afterObj) // needed after the tx lifetime
+			}
+			return err
+		}); err != nil {
+			return err
+		}
+		if fromBkt == nil {
+			break
+		}
 	}
-	return updateVersion(tx, 4)
+	return db.Update(func(tx *bbolt.Tx) error { return updateVersion(tx, 4) })
 }
 
-func migrateContainersToMetaBucket(l *zap.Logger, cs Containers, tx *bbolt.Tx) error {
+func migrateContainersToMetaBucket(l *zap.Logger, cs Containers, tx *bbolt.Tx, fromBkt, afterObj []byte) ([]byte, []byte, error) {
 	c := tx.Cursor()
-	for name, _ := c.First(); name != nil; name, _ = c.Next() {
+	var name []byte
+	if fromBkt != nil {
+		name, _ = c.Seek(fromBkt)
+	} else {
+		name, _ = c.First()
+	}
+	rem := uint(1000)
+	var done uint
+	var err error
+	for ; name != nil; name, _ = c.Next() {
 		switch name[0] {
 		default:
 			continue
 		case primaryPrefix, tombstonePrefix, storageGroupPrefix, lockersPrefix, linkObjectsPrefix:
 		}
 		if len(name[1:]) != cid.Size {
-			return fmt.Errorf("invalid container bucket with prefix 0x%X: wrong CID len %d", name[0], len(name[1:]))
+			return nil, nil, fmt.Errorf("invalid container bucket with prefix 0x%X: wrong CID len %d", name[0], len(name[1:]))
 		}
 		cnr := cid.ID(name[1:])
 		if exists, err := cs.Exists(cnr); err != nil {
-			return fmt.Errorf("check container presence: %w", err)
+			return nil, nil, fmt.Errorf("check container presence: %w", err)
 		} else if !exists {
 			l.Info("container no longer exists, ignoring", zap.Stringer("container", cnr))
 			continue
 		}
 		b := tx.Bucket(name) // must not be nil, bbolt/Tx.ForEach follows the same assumption
-		if err := migrateContainerToMetaBucket(l, tx, b.Cursor(), cnr); err != nil {
-			return fmt.Errorf("process container 0x%X%s bucket: %w", name[0], cnr, err)
+		if done, afterObj, err = migrateContainerToMetaBucket(l, tx, b.Cursor(), cnr, afterObj, rem); err != nil {
+			return nil, nil, fmt.Errorf("process container 0x%X%s bucket: %w", name[0], cnr, err)
+		}
+		if done == rem {
+			break
 		}
+		rem -= done
 	}
-	return nil
+	return name, afterObj, nil
 }
 
-func migrateContainerToMetaBucket(l *zap.Logger, tx *bbolt.Tx, c *bbolt.Cursor, cnr cid.ID) error {
-	for k, v := c.First(); k != nil; k, v = c.Next() {
-		if err := migrateObjectToMetaBucket(l, tx, cnr, k, v); err != nil {
-			return err
+func migrateContainerToMetaBucket(l *zap.Logger, tx *bbolt.Tx, c *bbolt.Cursor, cnr cid.ID, after []byte, limit uint) (uint, []byte, error) {
+	var k, v []byte
+	if after != nil {
+		if k, v = c.Seek(after); bytes.Equal(k, after) {
+			k, v = c.Next()
 		}
+	} else {
+		k, v = c.First()
 	}
-	return nil
+	var done uint
+	for ; k != nil; k, v = c.Next() {
+		ok, err := migrateObjectToMetaBucket(l, tx, cnr, k, v)
+		if err != nil {
+			return 0, nil, err
+		}
+		if ok {
+			if done++; done == limit {
```

Comment:
done++
if done == limit

Do not obfuscate things.

Reply:
this is a valid Go code

Reply:
A lot of strings can be valid Go code, it doesn't mean they're all good. Mixing conditions and state changes is a classic problem, too easy to overlook when doing something in the same codebase a year after.

Reply:
> Mixing conditions and state changes is a classic problem

Go authors do not recommend using it?

Reply:
It's not related to Go, this is relevant for any structural language. And yes `err := smth(); err !=nil` is not the best thing too, it's just it's so common in Go that we're used to it (`err` locality can tiny bit helpful sometimes as well).

But this has the point: https://www.teamten.com/lawrence/programming/keep-if-clauses-side-effect-free.html

And these guys: https://web.git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/coding-style.rst?h=v6.14-rc5
know something too:

> Avoid tricky expressions.